plugins/storage/object/cloudian/src/main/java/org/apache/cloudstack/storage/datastore/driver/CloudianHyperStoreObjectStoreDriverImpl.java

```diff
@@ -0,0 +1,877 @@
+                for (AccessKeyMetadata accessKeyMetadata : listAccessKeyResult.getAccessKeyMetadata()) {
+                    if (iamAccessKeyId.equals(accessKeyMetadata.getAccessKeyId())) {
+                        return null;  // The IAM AccessKeyId still exists (as expected). return null.
+                    }
+                    // Usually, there will only be 1 credential that we manage, but an error persisting
+                    // credentials might leave an un-managed credential which we can just delete. It is better
+                    // to delete as otherwise, we may hit a max credential limit for this IAM user.
+                    DeleteAccessKeyRequest deleteAccessKeyRequest = new DeleteAccessKeyRequest();
+                    deleteAccessKeyRequest.setUserName(iamUser);
+                    deleteAccessKeyRequest.setAccessKeyId(accessKeyMetadata.getAccessKeyId());
+                    logger.info("Deleting un-managed IAM AccessKeyId {} for IAM User {}", accessKeyMetadata.getAccessKeyId(), iamUser);
+                    iamClient.deleteAccessKey(deleteAccessKeyRequest);
+                }
+            } catch (NoSuchEntityException e) {
+                // No IAM User. Ignore and fix this below.
+            }
+        }
+
+        // If we get here, a usable credential does not yet exist so create it.
+        // Before creating it, we also need to ensure the IAM User that will own it exists.
+        boolean createdUser = false;
+        try {
+            iamClient.createUser(new CreateUserRequest(iamUser));
+            logger.info("Created IAM user {} for account", iamUser);
+            createdUser = true;
+        } catch (EntityAlreadyExistsException e) {
+            // User already exists. Ignore and continue.
+        }
+
+        // Always Add or Update the IAM policy
+        iamClient.putUserPolicy(new PutUserPolicyRequest(iamUser, CloudianHyperStoreUtil.IAM_USER_POLICY_NAME, CloudianHyperStoreUtil.IAM_USER_POLICY));
+
+        if (! createdUser && iamAccessKeyId == null) {
+            // User already exists but we never saved any access key before. We should try clean up
+            logger.debug("Looking for any un-managed IAM credentials for IAM User {}", iamUser);
+            ListAccessKeysResult listRes = iamClient.listAccessKeys(new ListAccessKeysRequest().withUserName(iamUser));
+            for (AccessKeyMetadata accessKeyMetadata : listRes.getAccessKeyMetadata()) {
+                DeleteAccessKeyRequest deleteAccessKeyRequest = new DeleteAccessKeyRequest();
+                deleteAccessKeyRequest.setUserName(iamUser);
+                deleteAccessKeyRequest.setAccessKeyId(accessKeyMetadata.getAccessKeyId());
+                logger.info("Deleting un-managed IAM AccessKeyId {} for IAM User {}", accessKeyMetadata.getAccessKeyId(), iamUser);
+                iamClient.deleteAccessKey(deleteAccessKeyRequest);
+            }
```

Comment:
Same here, should we be cleaning up stuff on a method used to create credentials?

Reply:
Thanks yes. Same answer really. The plugin should manage this for the account that is created specifically for CloudStack.